solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java

```diff
@@ -146,14 +158,82 @@ public SolrQueryRequest getRequest() {
   @Override
   public ModelWeight createWeight(IndexSearcher searcher, boolean needsScores, float boost)
       throws IOException {
-    final FeatureWeight[] allFeatureWeights = createWeights(meta.getAllFeatures(),
-        searcher, needsScores);
-    final FeatureWeight[] modelFeaturesWeights = createWeights(meta.getFeatures(),
-        searcher, needsScores);
-
-    return new ModelWeight(searcher, modelFeaturesWeights, allFeatureWeights);
+    boolean newCode = true;
+    if (!newCode){
+      long  start = System.currentTimeMillis();
+      final FeatureWeight[] allFeatureWeights = createWeights(meta.getAllFeatures(),
+          searcher, needsScores);
+      final FeatureWeight[] modelFeaturesWeights = createWeights(meta.getFeatures(),
+          searcher, needsScores);
+      long  end = System.currentTimeMillis();
+      log.info("In ModelQuery.java::createWeight() OLD createWeight time:{} ms", (end-start));
+      return new ModelWeight(searcher, modelFeaturesWeights, allFeatureWeights);
+    }
+    else{      
+      long  start = System.currentTimeMillis();
+      final Collection<Feature> modelFeatures = meta.getFeatures();
+      final Collection<Feature> allFeatures = meta.getAllFeatures();
+          
+      int modelFeatSize = modelFeatures.size();
+      int allFeatSize = this.featuresRequested ? allFeatures.size() : modelFeatSize;
+
+      final FeatureWeight[] allFeatureWeights = new FeatureWeight[allFeatSize];
+      final FeatureWeight[] modelFeaturesWeights = new FeatureWeight[modelFeatSize];
+      createWeightsSelectively(allFeatures,modelFeatures,searcher, needsScores, allFeatureWeights,modelFeaturesWeights); 
+      
+      HashMap<Integer, Normalizer> modelFeatureNorms = new HashMap<Integer, Normalizer>(); // store the featureNorms for modelFeatures to use later for normalization
```

Comment:
Should this not just be an array that maps 1 to 1 with the modelFeaturesWeights array to index into with featureId?